src/controller/proxy/local.go

```diff
@@ -0,0 +1,199 @@
+	return l.registry.BlobExist(art.Repository, art.Digest)
+}
+
+func (l *localHelper) init() {
+	if l.registry != nil {
+		return
+	}
+	log.Debugf("core url:%s, local core url: %v", config.GetCoreURL(), config.LocalCoreURL())
+	// the traffic is internal only
+	registryURL := config.LocalCoreURL()
+	authorizer := comHttpAuth.NewSecretAuthorizer(config.ProxyServiceSecret)
+	l.registry = registry.NewClientWithAuthorizer(registryURL, authorizer, true)
+}
+
+func (l *localHelper) PushBlob(localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error {
+	log.Debugf("Put blob to local registry, localRepo:%v, digest: %v", localRepo, desc.Digest)
+	ref := string(desc.Digest)
+	artName := localRepo + ":" + ref
+	// use inflight checker to avoid multiple requests to push blob to local in same time
+	if !inflightChecker.addRequest(artName) {
+		return nil
+	}
+	defer inflightChecker.removeRequest(artName)
+	err := l.registry.PushBlob(localRepo, ref, desc.Size, bReader)
+	return err
+}
+
+func (l *localHelper) PushManifest(repo string, ref string, manifest distribution.Manifest) error {
+	// Make sure there is only one go routing to push current artName to local repo
+	artName := repo + ":" + ref
+	// use inflight checker to avoid multiple requests to push manifest to local in same time
+	if !inflightChecker.addRequest(artName) {
+		return nil
+	}
+	defer inflightChecker.removeRequest(artName)
+
+	mediaType, payload, err := manifest.Payload()
+	if err != nil {
+		return err
+	}
+	_, err = l.registry.PushManifest(repo, ref, mediaType, payload)
+	return err
+}
+
+// DeleteManifest cleanup delete tag from local repo
+func (l *localHelper) DeleteManifest(repo, ref string) {
+	log.Debug("Remove tag from repo if it is exist")
+	if err := l.registry.DeleteManifest(repo, ref); err != nil {
+		// sometimes user pull a non-exist image
+		log.Warningf("failed to remove artifact, error %v", err)
+	}
+}
+
+// updateManifestList -- Trim the manifest list, make sure at least one depend manifests is ready
+func (l *localHelper) updateManifestList(ctx context.Context, repo string, manifest distribution.Manifest) (distribution.Manifest, error) {
+	switch v := manifest.(type) {
+	case *manifestlist.DeserializedManifestList:
+		existMans := make([]manifestlist.ManifestDescriptor, 0)
+		for _, m := range v.Manifests {
+			art := lib.ArtifactInfo{Repository: repo, Digest: string(m.Digest)}
+			exist, err := l.ManifestExist(ctx, art)
+			if err != nil {
+				continue
+			}
+			if exist {
+				existMans = append(existMans, m)
+			}
+		}
+		if len(existMans) == 0 {
+			return nil, errManifestNotFound
+		}
+		return manifestlist.FromDescriptors(existMans)
+	}
+	log.Errorf("Current manifest list type is unknown, manifest type[%T], content [%+v]", manifest, manifest)
+	return manifest, nil
+}
+
+func (l *localHelper) PushManifestList(ctx context.Context, repo string, ref string, man distribution.Manifest) error {
```

Comment:
It should try best to wait till all manifests referenced by the manifest list is pulled.

Reply:
Updated